server/neptune/workflows/internal/pr/revision/policy/handler.go

```diff
@@ -0,0 +1,101 @@
+package policy
+
+import (
+	"github.com/runatlantis/atlantis/server/neptune/workflows/activities"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/pr/revision"
+	temporalInternal "github.com/runatlantis/atlantis/server/neptune/workflows/internal/temporal"
+	"go.temporal.io/sdk/workflow"
+	"time"
+)
+
+type dismisser interface {
+	Dismiss(ctx workflow.Context, revision revision.Revision) error
+}
+
+type policyFilter interface {
+	Filter(ctx workflow.Context, revision revision.Revision, failedPolicies []activities.PolicySet) ([]activities.PolicySet, error)
+}
+
+type NewApproveSignalRequest struct{}
+
+type FailedPolicyHandler struct {
+	ApprovalSignalChannel workflow.ReceiveChannel
+	Dismisser             dismisser
+	PolicyFilter          policyFilter
+}
+
+type Action int64
+
+const (
+	onApprovalSignal Action = iota
+	onPollTick
+)
+
+func (f *FailedPolicyHandler) Handle(ctx workflow.Context, revision revision.Revision, failedPolicies []activities.PolicySet) {
+	policies := dedup(failedPolicies)
+
+	var action Action
+	s := temporalInternal.SelectorWithTimeout{
+		Selector: workflow.NewSelector(ctx),
+	}
+	s.AddReceive(f.ApprovalSignalChannel, func(c workflow.ReceiveChannel, more bool) {
+		defer func() {
+			action = onApprovalSignal
+		}()
+		if !more {
+			return
+		}
+		var request NewApproveSignalRequest
+		c.Receive(ctx, &request)
+	})
+	onTimeout := func(f workflow.Future) {
+		_ = f.Get(ctx, nil)
+		action = onPollTick
+	}
+	cancelTimer, _ := s.AddTimeout(ctx, 5*time.Minute, onTimeout)
+
+	for {
+		if len(policies) == 0 {
+			break
+		}
+		s.Select(ctx)
+		if action == onPollTick {
+			cancelTimer()
+			cancelTimer, _ = s.AddTimeout(ctx, 10*time.Minute, onTimeout)
```

Comment:
Seems awfully short, also what's going to happen in this case? a customer is forced to run plan again?  What if we just don't time this out ever and instead just have the main timeout cancel this if we need to

Reply:
The timeout doesn't end the PolicyHandler, it serves as our poll tick to retry fetching GH PR state for approvals. The only ways this handler shuts down is if we cancel the ctx because the parent revision processor has received a new revision to work on or we've met are goal case of no failing policies left.


`awfully short` --> yeah with approvals being so user dependent, I'm not sure if there is a good time to poll that is both short enough that when a user is waiting for it, it won't take too long; but long enough that it won't stress our GH ratelimit and increase load on our system.